### PR TITLE
Modernize null pointers to use nullptr

### PR DESCRIPTION
**Pull request by masterleinad**
This PR replaces all null pointers by nullptr. It passes a full test suite run and the gcc warning `-Wzero-as-null-pointer-constant` apart from the files in the bundled packages.

Mostly generated by `run-clang-tidy.py -checks=modernize-use-nullptr -fix`

### CODE REVIEW

**Comment by bangerth** [comment]
Can you fix the indentation issue?

File: include/deal.II/base/array_view.h [hunk 1]
```diff
@@ -157,7 +157,7 @@ template <typename ElementType>
 inline
 ArrayView<ElementType>::ArrayView()
   :
-  starting_element (NULL),
+  starting_element ((nullptr)),
```
Comment (bangerth) [inline include/deal.II/base/array_view.h]: Do we need the second pair of parentheses?

File: include/deal.II/base/index_set.h [hunk 2]
```diff
@@ -1001,7 +1001,7 @@ IndexSet::IntervalIterator::IntervalIterator(const IndexSet *idxset, const Index
 
 inline
 IndexSet::IntervalIterator::IntervalIterator()
-  : accessor(NULL)
+  : accessor((nullptr))
```
Comment (bangerth) [inline include/deal.II/base/index_set.h]: same here

File: include/deal.II/base/memory_consumption.h [hunk 3]
```diff
@@ -280,7 +280,7 @@ namespace MemoryConsumption
   inline
   std::size_t memory_consumption (const char *string)
   {
-    if (string == NULL)
+    if (string == (nullptr))
```
Comment (bangerth) [inline include/deal.II/base/memory_consumption.h]: and here

File: include/deal.II/base/parameter_handler.h [hunk 4]
```diff
@@ -288,8 +288,8 @@ namespace Patterns
     /**
      * Creates a new object on the heap using @p new if the given
      * @p description is a valid format (for example created by calling
-     * description() on an existing object), or NULL otherwise. Ownership of
-     * the returned object is transferred to the caller of this function,
+     * description() on an existing object), or nullptr otherwise. Ownership
```
Comment (bangerth) [inline include/deal.II/base/parameter_handler.h]: can you mark this up as `@p nullptr`?

File: include/deal.II/base/thread_management.h [hunk 5]
```diff
@@ -2902,7 +2902,7 @@ namespace Threads
     TaskDescriptor<RT>::TaskDescriptor (const std::function<RT ()> &function)
       :
       function (function),
-      task(NULL),
+      task((nullptr)),
```
Comment (bangerth) [inline include/deal.II/base/thread_management.h]: parens

File: include/deal.II/base/time_stepping.templates.h [hunk 6]
```diff
@@ -415,7 +415,7 @@ namespace TimeStepping
     refine_tol(refine_tol),
     coarsen_tol(coarsen_tol),
     last_same_as_first(false),
-    last_stage(NULL)
+    last_stage((nullptr))
```
Comment (bangerth) [inline include/deal.II/base/time_stepping.templates.h]: parens

File: include/deal.II/base/time_stepping.templates.h [hunk 7]
```diff
@@ -665,10 +665,10 @@ namespace TimeStepping
   template <typename VectorType>
   void EmbeddedExplicitRungeKutta<VectorType>::free_memory()
   {
-    if (last_stage!=NULL)
+    if (last_stage!=(nullptr))
```
Comment (bangerth) [inline include/deal.II/base/time_stepping.templates.h]: parens

File: include/deal.II/base/time_stepping.templates.h [hunk 8]
```diff
@@ -665,10 +665,10 @@ namespace TimeStepping
...
       delete last_stage;
 
-    last_stage = NULL;
+    last_stage = (nullptr);
```
Comment (bangerth) [inline include/deal.II/base/time_stepping.templates.h]: parens

File: include/deal.II/base/time_stepping.templates.h [hunk 9]
```diff
@@ -762,7 +762,7 @@ namespace TimeStepping
     // Save the last stage if necessary
     if (last_same_as_first==true)
       {
-        if (last_stage==NULL)
+        if (last_stage==(nullptr))
```
Comment (bangerth) [inline include/deal.II/base/time_stepping.templates.h]: parens

File: include/deal.II/base/time_stepping.templates.h [hunk 10]
```diff
@@ -817,7 +817,7 @@ namespace TimeStepping
     // of the first stage.
     if (last_same_as_first==true)
       {
-        if (last_stage!=NULL)
+        if (last_stage!=(nullptr))
```
Comment (bangerth) [inline include/deal.II/base/time_stepping.templates.h]: parens

File: include/deal.II/dofs/dof_accessor.templates.h [hunk 11]
```diff
@@ -60,7 +60,7 @@ DoFAccessor (const Triangulation<DoFHandlerType::dimension,DoFHandlerType::space
                                                index),
   dof_handler(const_cast<DoFHandlerType *>(dof_handler))
 {
-  Assert (tria == NULL || &dof_handler->get_triangulation() == tria,
+  Assert (tria ==nullptr|| &dof_handler->get_triangulation() == tria,
```
Comment (bangerth) [inline include/deal.II/dofs/dof_accessor.templates.h]: keep the spaces on either side of `nullptr`.

File: include/deal.II/fe/fe_enriched.h [hunk 12]
```diff
@@ -136,7 +136,7 @@ DEAL_II_NAMESPACE_OPEN
  * @code
  *   FE_Enriched<dim> fe_non_enriched(FE_Q<dim>(1),
  *                                    FE_Nothing<dim>(1,true),
- *                                    NULL);
+ *                                    (nullptr));
```
Comment (bangerth) [inline include/deal.II/fe/fe_enriched.h]: parens

File: include/deal.II/grid/tria.h [hunk 13]
```diff
@@ -934,7 +934,7 @@ namespace internal
  * classes. Again, see
  * @ref GlossUserData "the glossary for more information".
  *
- * The value of these user indices or pointers is @p NULL by default. Note
+ * The value of these user indices or pointers is @pnullptrby default. Note
```
Comment (bangerth) [inline include/deal.II/grid/tria.h]: spaces missing

File: include/deal.II/grid/tria_accessor.h [hunk 14]
```diff
@@ -1104,7 +1104,7 @@ class TriaAccessor : public TriaAccessorBase<structdim, dim, spacedim>
   void set_user_pointer (void *p) const;
 
   /**
-   * Reset the user pointer to a @p NULL pointer. See
+   * Reset the user pointer to a @pnullptrpointer. See
```
Comment (bangerth) [inline include/deal.II/grid/tria_accessor.h]: spaces

File: include/deal.II/grid/tria_accessor.templates.h [hunk 15]
```diff
@@ -2134,7 +2134,7 @@ inline
 TriaAccessor<0, dim, spacedim>::
 TriaAccessor (const TriaAccessor<structdim2,dim2,spacedim2> &)
   :
-  tria (NULL),
+  tria ((nullptr)),
```
Comment (bangerth) [inline include/deal.II/grid/tria_accessor.templates.h]: parens

File: include/deal.II/grid/tria_accessor.templates.h [hunk 16]
```diff
@@ -2148,7 +2148,7 @@ inline
 TriaAccessor<0, dim, spacedim>::
 TriaAccessor (const InvalidAccessor<structdim2,dim2,spacedim2> &)
   :
-  tria (NULL),
+  tria ((nullptr)),
```
Comment (bangerth) [inline include/deal.II/grid/tria_accessor.templates.h]: parens

File: include/deal.II/lac/la_parallel_vector.templates.h [hunk 17]
```diff
@@ -831,7 +831,7 @@ namespace LinearAlgebra
       // If no communication pattern is given, create one. Otherwise, use the
       // given one.
       std::shared_ptr<const Utilities::MPI::Partitioner> comm_pattern;
-      if (communication_pattern.get() == NULL)
+      if (communication_pattern.get() == (nullptr))
```
Comment (bangerth) [inline include/deal.II/lac/la_parallel_vector.templates.h]: parens

File: include/deal.II/lac/petsc_compatibility.h [hunk 18]
```diff
@@ -53,7 +53,7 @@ namespace PETScWrappers
 #if DEAL_II_PETSC_VERSION_LT(3, 7, 0)
     const PetscErrorCode ierr = PetscOptionsSetValue (name.c_str (), value.c_str ());
 #else
-    const PetscErrorCode ierr = PetscOptionsSetValue (NULL, name.c_str (), value.c_str ());
+    const PetscErrorCode ierr = PetscOptionsSetValue ((nullptr), name.c_str (), value.c_str ());
```
Comment (bangerth) [inline include/deal.II/lac/petsc_compatibility.h]: There's an issue here with parentheses, but that's also a call to a C function -- leaving `NULL` here may be appropriate, or maybe not. I don't think I care one way or another.

File: include/deal.II/lac/petsc_compatibility.h [hunk 19]
```diff
@@ -72,7 +72,7 @@ namespace PETScWrappers
    */
   inline PetscErrorCode destroy_matrix (Mat &matrix)
   {
-    // PETSc will check whether or not matrix is NULL.
+    // PETSc will check whether or not matrix is (nullptr).
```
Comment (bangerth) [inline include/deal.II/lac/petsc_compatibility.h]: parens

File: include/deal.II/lac/petsc_compatibility.h [hunk 20]
```diff
@@ -94,7 +94,7 @@ namespace PETScWrappers
    */
   inline PetscErrorCode destroy_krylov_solver (KSP &krylov_solver)
   {
-    // PETSc will check whether or not matrix is NULL.
+    // PETSc will check whether or not matrix is (nullptr).
```
Comment (bangerth) [inline include/deal.II/lac/petsc_compatibility.h]: parens

File: include/deal.II/lac/petsc_vector_base.h [hunk 21]
```diff
@@ -1190,7 +1190,7 @@ namespace PETScWrappers
         PetscErrorCode ierr = VecGetOwnershipRange (vector, &begin, &end);
         AssertThrow (ierr == 0, ExcPETScError(ierr));
 
-        Vec locally_stored_elements = PETSC_NULL;
+        Vec locally_stored_elements = nullptr;
```
Comment (bangerth) [inline include/deal.II/lac/petsc_vector_base.h]: also questionable. I don't know whether we want to keep the PETSc convention here.

File: include/deal.II/lac/precondition.h [hunk 22]
```diff
@@ -1912,10 +1912,10 @@ namespace internal
                               std::shared_ptr<DiagonalMatrix<VectorType> > &preconditioner,
                               VectorType                                         &diagonal_inverse)
     {
-      if (preconditioner.get() == NULL ||
+      if (preconditioner.get() ==nullptr||
           preconditioner->m() != matrix.m())
         {
-          if (preconditioner.get() == NULL)
+          if (preconditioner.get() == (nullptr))
```
Comment (bangerth) [inline include/deal.II/lac/precondition.h]: parens

File: include/deal.II/lac/read_write_vector.h [hunk 23]
```diff
@@ -624,7 +624,7 @@ namespace LinearAlgebra
   inline
   ReadWriteVector<Number>::ReadWriteVector ()
     :
-    val(NULL)
+    val((nullptr))
```
Comment (bangerth) [inline include/deal.II/lac/read_write_vector.h]: parens

File: include/deal.II/lac/read_write_vector.h [hunk 24]
```diff
@@ -636,7 +636,7 @@ namespace LinearAlgebra
   ReadWriteVector<Number>::ReadWriteVector (const ReadWriteVector<Number> &v)
     :
     Subscriptor(),
-    val(NULL)
+    val((nullptr))
```
Comment (bangerth) [inline include/deal.II/lac/read_write_vector.h]: parens

File: include/deal.II/lac/read_write_vector.h [hunk 25]
```diff
@@ -647,7 +647,7 @@ namespace LinearAlgebra
   inline
   ReadWriteVector<Number>::ReadWriteVector (const size_type size)
     :
-    val(NULL)
+    val((nullptr))
```
Comment (bangerth) [inline include/deal.II/lac/read_write_vector.h]: parens

File: include/deal.II/lac/read_write_vector.h [hunk 26]
```diff
@@ -658,7 +658,7 @@ namespace LinearAlgebra
   inline
   ReadWriteVector<Number>::ReadWriteVector (const IndexSet &locally_stored_indices)
     :
-    val(NULL)
+    val((nullptr))
```
Comment (bangerth) [inline include/deal.II/lac/read_write_vector.h]: parens

File: include/deal.II/lac/read_write_vector.templates.h [hunk 27]
```diff
@@ -52,14 +52,14 @@ namespace LinearAlgebra
   {
     if (new_alloc_size == 0)
       {
-        if (val != NULL)
+        if (val != (nullptr))
```
Comment (bangerth) [inline include/deal.II/lac/read_write_vector.templates.h]: parens

File: include/deal.II/lac/read_write_vector.templates.h [hunk 28]
```diff
@@ -52,14 +52,14 @@ namespace LinearAlgebra
...
           free(val);
-        val = NULL;
+        val = (nullptr);
```
Comment (bangerth) [inline include/deal.II/lac/read_write_vector.templates.h]: parens

File: include/deal.II/lac/read_write_vector.templates.h [hunk 29]
```diff
@@ -52,14 +52,14 @@ namespace LinearAlgebra
...
         thread_loop_partitioner.reset(new parallel::internal::TBBPartitioner());
       }
     else
       {
-        if (val != NULL)
+        if (val != (nullptr))
```
Comment (bangerth) [inline include/deal.II/lac/read_write_vector.templates.h]: parens

File: include/deal.II/lac/read_write_vector.templates.h [hunk 30]
```diff
@@ -206,7 +206,7 @@ namespace LinearAlgebra
     // If no communication pattern is given, create one. Otherwise, use the
     // given one.
     std::shared_ptr<const Utilities::MPI::Partitioner> comm_pattern;
-    if (communication_pattern.get() == NULL)
+    if (communication_pattern.get() == (nullptr))
```
Comment (bangerth) [inline include/deal.II/lac/read_write_vector.templates.h]: parens

File: include/deal.II/lac/read_write_vector.templates.h [hunk 31]
```diff
@@ -305,7 +305,7 @@ namespace LinearAlgebra
 
     // If no communication pattern is given, create one. Otherwise, use the one
     // given.
-    if (communication_pattern == NULL)
+    if (communication_pattern == (nullptr))
```
Comment (bangerth) [inline include/deal.II/lac/read_write_vector.templates.h]: parens

File: include/deal.II/lac/read_write_vector.templates.h [hunk 32]
```diff
@@ -315,7 +315,7 @@ namespace LinearAlgebra
           {
             epetra_comm_pattern =
               std::dynamic_pointer_cast<const EpetraWrappers::CommunicationPattern> (comm_pattern);
-            if (epetra_comm_pattern == NULL)
+            if (epetra_comm_pattern == (nullptr))
```
Comment (bangerth) [inline include/deal.II/lac/read_write_vector.templates.h]: parens

File: include/deal.II/lac/relaxation_block.h [hunk 33]
```diff
@@ -91,7 +91,7 @@ class RelaxationBlock :
                     const typename PreconditionBlockBase<InverseNumberType>::Inversion inversion
                     = PreconditionBlockBase<InverseNumberType>::gauss_jordan,
                     const double threshold = 0.,
-                    VectorType *temp_ghost_vector = NULL);
+                    VectorType *temp_ghost_vector = (nullptr));
```
Comment (bangerth) [inline include/deal.II/lac/relaxation_block.h]: parens

File: include/deal.II/lac/solver_bicgstab.h [hunk 34]
```diff
@@ -296,15 +296,15 @@ SolverBicgstab<VectorType>::SolverBicgstab (SolverControl        &cn,
...
-  Vt(NULL),
-  Vv(NULL),
-  Vb(NULL),
+  Vx((nullptr)),
+  Vr((nullptr)),
+  Vrbar((nullptr)),
+  Vp((nullptr)),
+  Vy((nullptr)),
+  Vz((nullptr)),
+  Vt((nullptr)),
```
Comment (bangerth) [inline include/deal.II/lac/solver_bicgstab.h]: parens for all of these variables

File: include/deal.II/lac/solver_cg.h [hunk 35]
```diff
@@ -327,9 +327,9 @@ SolverCG<VectorType>::SolverCG (SolverControl        &cn,
                                 const AdditionalData     &data)
   :
   Solver<VectorType>(cn,mem),
-  Vr(NULL),
-  Vp(NULL),
-  Vz(NULL),
+  Vr((nullptr)),
+  Vp((nullptr)),
+  Vz((nullptr)),
```
Comment (bangerth) [inline include/deal.II/lac/solver_cg.h]: parens

File: include/deal.II/lac/petsc_compatibility.h (hunk 18 above)
Comment (drwells) [inline include/deal.II/lac/petsc_compatibility.h]: I have called SLEPc functions with `nullptr` and things work correctly, so this is fine (aside from the extra parentheses).

File: include/deal.II/lac/petsc_vector_base.h (hunk 21 above)
Comment (drwells) [inline include/deal.II/lac/petsc_vector_base.h]: Formally, PETSC_NULL is deprecated and they say we should use NULL, so I believe nullptr would be okay here too.

**Comment by masterleinad** [comment]
> I made it down to the solver_cg.h file and then fell asleep -- what a boring patch :-)

Yeah, sorry, no fancy lambdas. :wink:

> There are a lot of places where you unnecessarily put parentheses around nullptr. Can you fix the ones I marked up, and see if you can find others (likely) further down in the patch?

Thanks for pointing out. I fixed everything you marked using regex.

I was also not sure which style would be preferred for the PETSc/SLEPc calls, so I put them in a separate commit. The tests clearly also pass (for me) using `nullptr` instead of `PETSC_NULL`.

**Comment by bangerth** [comment]
On 04/07/2017 04:29 AM, Daniel Arndt wrote:
> I was also not sure which style would be preferred for the PETSc/SLEPc
> calls, so I put them in a separate commit. The tests clearly also pass
> (for me) using |nullptr| instead of |PETSC_NULL|.

I didn't mean to suggest that that is wrong from a language perspective. 
More a style issue -- that we ought to call PETSc functions with 
arguments the way they are typically written.

But that has the disadvantage that you still get warnings, so let's do 
`nullptr` as you suggest.

File: include/deal.II/matrix_free/operators.h [hunk 36]
```diff
@@ -882,19 +882,15 @@ namespace MatrixFreeOperators
     :
     Subscriptor(),
     have_interface_matrices(false)
-  {
-    // boost-1.62.0 doesn't allow initializing a shared_ptr
-    // with NULL. Make sure the default constructor does that.
-    Assert(data.get() == NULL, ExcInternalError());
-  }
+  {}
```
Comment (bangerth) [inline include/deal.II/matrix_free/operators.h]: Yes :-)

**Comment by bangerth** [comment]
OK to merge once the tester is happy with it.

I'm going to note that there will be conflicts with some of @tamiko 's patches removing files and classes. I don't know whose pain would be greater to resolve these conflicts, so I'm going to merge patches in the order in which they are approved by the tester.

**Comment by masterleinad** [comment]
Rebased

**Pull request merged**